### PR TITLE
Fix the broken ImageMagick loader

### LINKED ISSUE #13090: The ImageMagick image loader is unusable
**Describe the bug/issue**

If you're trying to import images in formats that the IM loader handles (meaning you have to build a dt with the IM library for that, not GM), you're going to run into problems. With some image formats, dt crashes immediately upon import, with others the import is successful, but dt hangs on exit.

**To Reproduce**
1. Build dt with ImageMagick enabled
2. Get images in formats supported by IM (but not by other dt loaders, so they are processed by the IM loader). These can be, for example, such formats as dcm and miff.
3. Try importing these images.

In the case of miff images (at least the ones I used), dt immediately crashes. In the case of dcm images dt hangs on exit.

Test images:
http://www.dclunie.com/images/FluoroWithDisplayShutter.dcm.zip
https://github.com/dangom/sample-dicom/blob/master/MR000000.dcm
https://github.com/ImageMagick/ImageMagick/tree/main/tests

**Expected behavior**

dt should import images correctly, without hangs and crashes.

**Which commit introduced the error**

This has not been investigated yet.

**Platform**

* darktable version : Latest git master build as of the date of the issue.
* OS  : Windows 10

### PR DESCRIPTION
**Pull request by victoryforce**
Resolves #13090. We should not call MagicWandTerminus() in case of successful image reading.

### CODE REVIEW

File: src/common/imageio_im.c [hunk 1]
```diff
@@ -126,8 +126,10 @@ dt_imageio_retval_t dt_imageio_open_im(dt_image_t *img, const char *filename, dt
     MagickRelinquishMemory(profile_data);
   }
 
+  // As a warning to those who will modify the loader in the future:
+  // MagickWandTerminus() cannot be called on successful image reading.
+  // See https://github.com/darktable-org/darktable/issues/13090 regarding the consequences.
   DestroyMagickWand(image);
-  MagickWandTerminus();
```
Comment (kmilos) [inline src/common/imageio_im.c:130]: If so, should also then remove the matching `MagickWandGenesis()` at the start to avoid it creating a new environment on every file open?

Isn't a full fix to call matching `MagickWandGenesis()`/`MagickWandTerminus()` at the start/end of dt session then?

Comment (victoryforce) [reply src/common/imageio_im.c:130]: > If so, should also then remove the matching MagickWandGenesis() at the start to avoid it creating a new environment on every file open?

You added calls to MagicWandGenesis()/MagickWandTerminus() functions in 6f5e4980. Before this, the loader was working fine without these calls, although it's hard for me to say if such usage is correct/optimal. In any case, ImageMagick function calls worked even without creating an environment (whatever that term means).

In summary, the code works both without a MagicWandGenesis call and with a MagicWandGenesis, but without the corresponding explicit MagickWandTerminus.

I don't know which of these options is "more correct".

> Isn't a full fix to call matching MagickWandGenesis()/MagickWandTerminus() at the start/end of dt session then?

Even when we won't use ImageMagick in this session?

Comment (kmilos) [reply src/common/imageio_im.c:130]: Yes, that's why I sad we should remove them together as a matching pair. I just followed the IM API usage example, I'm not sure what they do exactly either, sorry 😯

**Comment by kmilos** [comment]
LGTM

**Comment by victoryforce** [comment]
@TurboGit Will you include this crash fix in 4.2?

**Comment by TurboGit** [comment]
@victoryforce : I don't see this as a regression as IM is not active by default anyway. So I suppose waiting for 4.4 won't hurt.

**Comment by victoryforce** [comment]
> @victoryforce : I don't see this as a regression as IM is not active by default anyway. So I suppose waiting for 4.4 won't hurt.

Agreed, this is definitely the case NOW, as IM is not currently used by default and probably won't be active in the builds of a large number of distro repos, not just our official builds.

But I had the following development path in mind:

- Make sure the IM loader is working again. Before this, the problems were first due to the incorrect build of the IM library in MinGW (it was already fixed, see [MinGW packages issue](https://github.com/msys2/MINGW-packages/issues/11777)), and then due to the bug that this PR fixes. This fix reverts one code change that was recently added and turned out to be buggy, so it just reverts the code to a state that was known to work, nothing more.

- Change the default and build dt with IM (and only if this library is not available, then with GM). The fact is that GM provides worse support for image formats:
    - with the support of some formats, it lags behind IM
    - where the format is supported by both libraries, GM can output images with gamut clipping, unlike IM (quoting a comment in the source code of GM: "Better to read an image somewhat bogus then to error out." - an approach that is absolutely unacceptable for dt)

If we build dt 4.2 with IM by default, we will not only switch to a more active, live library with more devs (most of the work on GM is on the shoulders of one person), but we will also immediately have support for the QOI format in dt 4.2, which is already in IM, but not in GM.

**Comment by kmilos** [comment]
FYI, looks like major distros (Debian, Ubuntu, Fedora, etc.) are still behind on IM 6 (and so is MacPorts), while we require 7?

https://repology.org/project/imagemagick/versions
https://repology.org/project/graphicsmagick/versions

Indeed might a good idea to switch, just a question of right moment...

**Comment by victoryforce** [comment]
> FYI, looks like major distros (Debian, Ubuntu, Fedora, etc.) are still behind on IM 6 (and so is MacPorts), while we require 7?

Ha, we no longer require IM 7: https://github.com/darktable-org/darktable/pull/12600 😉

> Indeed might a good idea to switch, just a question of right moment...

To me this is a good moment, why wait months when the change is so easy and just switches us to a well-maintained library.

**Comment by kmilos** [comment]
Ah, forgot about that 👍

**Review by TurboGit — APPROVED** [review]
Ok, safe and very localized so let's have this in 4.2 won't hurt more than an already broken loader :) Thanks!

**Comment by victoryforce** [comment]
> P.S. It's not as easy as reordering the two find statements in CMake - at least for Windows, there is quite a bit more special packaging going on wrt GM for the installer, so better leave some time for testing... But it doesn't stop this particular PR 😉

Yes, you're right, it gets tricky for Windows. And, unfortunately, it is not enough to just include the IM library DLLs instead of those DLLs that are there now for the GM library.

This is starting to look like an issue that has a fairly low usage impact, but requires a disproportionate amount of research time relative to the improvement benefit.

After all, both libraries manage to support exotic formats with equal success. As for those formats where the GM does worse, in practice this should not be noticeable. For example, there is always an option to compile the dt directly with libjxl and use the native loader to avoid reading images with wide gamut clipped or compressed to sRGB.

**Comment by kmilos** [comment]
@victoryforce On Windows, indeed copying only the coder DLLs (like it's done for GM) doesn't work. But if I copy _everything_ (`etc`, `share`, ...) [in the package](https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-imagemagick?repo=ucrt64) over to the dt install directory, then it works. So just need to figure out what is the minimum required, we'll get there ;)
